File: 剑指offer/数组中出现次数超过一半的数字.cpp
#include <vector>
using namespace std;

class Solution {
public:
    int MoreThanHalfNum_Solution(vector<int> numbers) {
        int res = 0;
        int cnt = 0;
        for (int i = 0; i < numbers.size(); ++i) {
            if (!cnt) {
                res = numbers.at(i);
                cnt = 1;
            } else if (numbers.at(i) == res) {
                ++cnt;
            } else {
                --cnt;
            }
        }
        cnt = 0;
        for (int i = 0; i < numbers.size(); ++i) {
            if (numbers.at(i) == res)
                ++cnt;
        }
        if (cnt * 2 < numbers.size())   res = 0;
        return res;
    }
};

File: 剑指offer/数组中出现次数超过一半的数字.js
function MoreThanHalfNum_Solution(numbers)
{
  // write code here
  let cnt = 0
  let res
  for (let i = 0; i < numbers.length; ++i) {
    if (!cnt) res = numbers[i]
    cnt = numbers[i] === res ? cnt + 1 : cnt - 1
  }
  cnt = 0
  for (let i = 0; i < numbers.length; ++i) {
    cnt = numbers[i] === res ? cnt + 1 : cnt
  }
  if (cnt * 2 <= numbers.length)  res = 0
  return res
}
